quesma/quesma/auth_middleware.go

```diff
@@ -0,0 +1,69 @@
+// Copyright Quesma, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+package quesma
+
+import (
+	"net/http"
+	"quesma/elasticsearch"
+	"quesma/quesma/config"
+	"sync"
+	"time"
+)
+
+const cacheWipeInterval = 10 * time.Minute
+
+// authMiddleware a simple implementation of an authentication middleware,
+// which checks the Authorization header and validates it against Elasticsearch.
+//
+// If the validation is positive, the Authorization header is stored in a cache to avoid unnecessary calls to Elasticsearch preceding each request.
+// The cache is wiped every 10 minutes - all items at once, perhaps this could be revisited in the future.
+type authMiddleware struct {
+	nextHttpHandler   http.Handler
+	authHeaderCache   sync.Map
+	cacheWipeInterval time.Duration
+	esClient          elasticsearch.SimpleClient
+}
+
+func NewAuthMiddleware(next http.Handler, esConf config.ElasticsearchConfiguration) http.Handler {
+	esClient := elasticsearch.NewSimpleClient(&esConf)
+	middleware := &authMiddleware{nextHttpHandler: next, esClient: *esClient, cacheWipeInterval: cacheWipeInterval}
+	go middleware.startCacheWipeScheduler()
+	return middleware
+}
+
+func (a *authMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	auth := r.Header.Get("Authorization")
+	if auth == "" {
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+	if _, ok := a.authHeaderCache.Load(auth); ok {
+		a.nextHttpHandler.ServeHTTP(w, r)
+		return
+	}
+
+	if authenticated := a.esClient.Authenticate(auth); authenticated {
+		a.authHeaderCache.Store(auth, struct{}{})
```

Comment:
Why we don't cache `false` results?

Reply:
At first I was leaning towards more complex cache: storing both positive and negative attempts plus time of adding an item so that we wipe old entries based on that.

Then I had some afterthoughts and chosen this simpler implementation to save on PRs LoC :) 

Main purpose of this cache was to make Quesma's responses faster - even if someone will be hammering Elasticsearch with failed auth attempts - these will still be 401's anyways so I didn't see much of a value from caching those :)

Reply:
I don't know how elastic works. What happens when we're hammering with failed auths? Does Elastic locks that account? If yes, these requests are ours, even if we're proxy them.

Reply:
Nothing like that happens as far as I'm concerned within Elasticsearch.
External auth providers (e.g. SAML) might do such things, but that's a different thing :)